### PR TITLE
introduction of the first 64-bit Zephyr target

### PR DESCRIPTION
**Pull request by npitre**
~Note: this PR includes commits from PR #17442 as they're required for
a successful testsuite pass.~

This introduces a new target: native_posix_64. It is made separate from
native_posix so both the 32-bit and 64-bit variants can be tested in the
same CI run.

### CODE REVIEW

File: include/sys/util.h
Comment (marc-hb) [inline include/sys/util.h]: > large enough in practice.

How about "PTR_TO_INT_DIFF(x, y)" instead? This would consistent with the other macros above and  it would not miss the important "_INT_" part that guards the user against using it for a distance bigger than 2^sizeof(int).

Now it would be slightly longer to type but again not longer than the previous macros and to be really honest I'm not sure usage of a macro made with 90% casts (!) should be easy anyway... In fact the next commit shows this new macro is needed in only one (test) file across the entire Zephyr codebase, is that correct? If correct then this is probably not worth include/sys/util.h. Maybe some cast-free, (y-x)*sizeof() alternative local to that test file is enough.

**Comment by marc-hb** [comment]
> https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/45863/1/console
> Shippable — Run 45863 status is FAILED.

Only one of the 5 shards failed. It failed before having a chance to actually run any test with a segfault... of `scripts/kconfig/kconfig.py`! Closing and re-opening to see if it happens again.

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck --inline-logs -N +modified_tests.args --save-tests test_file.txt
JOBS: 6
Cleaning output directory /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out
Selecting all possible platforms per test case
Building testcase defconfigs...
Filtering test cases...
+ rm -f modified_tests.args modified_boards.args
+ /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/scripts/sanitycheck --inline-logs -N --save-tests test_file.txt
JOBS: 6
Cleaning output directory /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out
Selecting default platforms per test case
Building testcase defconfigs...
Could not build defconfig for native_posix_tests_kernel_fatal_kernel.common.stack_protection_config-sanitycheck
Zephyr version: 1.14.99
-- Found PythonInterp: /usr/bin/python3 (found suitable version "3.5.2", minimum required is "3.4") 
-- Selected BOARD native_posix
-- Found west: /usr/local/bin/west (found suitable version "0.5.7", minimum required is "0.5.6")
-- Loading /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/boards/posix/native_posix/native_posix.dts as base
-- Overlaying /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/dts/common/common.dts
CMake Error at /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/cmake/kconfig.cmake:200 (message):
  command failed with return code: Segmentation fault
Call Stack (most recent call first):
  /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/cmake/app/boilerplate.cmake:432 (include)
  CMakeLists.txt:4 (include)


-- Configuring incomplete, errors occurred!
Filtering test cases...
Couldn't build some defconfigs
+ exit 1
```

File: include/sys/util.h
Comment (pabigot) [reply include/sys/util.h]: Please see #17518.

**Comment by pabigot** [comment]
I checked this rebased on #17442 and it may have problems on Ubuntu 18:04 64-bit due to that distribution still using gcc 7.4 as the host compiler:
```
...
/usr/bin/ld: kernel/libkernel.a(device.c.obj): relocation R_X86_64_32S against `.rodata.config_levels.5657' can not be used when making a PIE object; recompile with -fPIC
/usr/bin/ld: kernel/libkernel.a(mempool.c.obj): relocation R_X86_64_32 against undefined symbol `_k_mem_pool_list_start' can not be used when making a PIE object; recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
```

I think we've had situations like this before.  I don't know what the workaround is.  I am using sdk 0.10.1.

File: boards/posix/native_posix/Kconfig.board
Comment (aescolar) [inline boards/posix/native_posix/Kconfig.board]: This is kind of ok. But I'm thinking it may confuse people, as one will now compile targetting native_posix or native_posix_64, but not native_posix_32. (mind you, I certainly think we should be "backwards compatible" and have "native_posix" still existing).
Really BOARD_NATIVE_POSIX_32BIT is nothing more than selecting BOARD_NATIVE_POSIX, or?

File: boards/posix/native_posix/Kconfig.board
Comment (aescolar) [inline boards/posix/native_posix/Kconfig.board]: As this is common for both, both it make sense to move it to BOARD_NATIVE_POSIX?

**Comment by aescolar** [comment]
CC @galak @nvlsianpu @wentongwu @ulfalizer (too many auto-reviewers due to too many files being changed)

File: boards/posix/native_posix/Kconfig.board
Comment (aescolar) [inline boards/posix/native_posix/Kconfig.board]: As this is common for both, would it make sense to move it to BOARD_NATIVE_POSIX?

File: arch/posix/CMakeLists.txt
Comment (aescolar) [inline arch/posix/CMakeLists.txt]: Yep. I think it would be nicer to add a new kconfig choice for the POSIX arch, something like `POSIX_ARCH_TARGET`
which could start with `POSIX_ARCH_TARGET{_X86,_X86_64}` and default to `POSIX_ARCH_TARGET_X86` if not `64BIT` (or have 64 bit one be selected by `NATIVE_POSIX_64BIT`).

Then this if below could be based on it instead of `64BIT`.

Whenever we start supporting other host targets we'd add new `POSIX_ARCH_TARGET`s and add new compile options here for them.

File: boards/posix/native_posix/Kconfig.defconfig
Comment (aescolar) [inline boards/posix/native_posix/Kconfig.defconfig]: wouldn't be better to say `if BOARD_NATIVE_POSIX_64BIT`?

**Review by aescolar — DISMISSED** [review]
(the review got submitted halfway, so it is split in 2)
* ~As the USB test patch already made it to master, please rebase without those commits to avoid confusion~(done)
* ~My main quarrel would be with the -m64, and the fPIC issue~(done). Otherwise I'm quite ok with this: the rest are mostly nits, or further improvements. 

I think it would be nice if @ulfalizer could give his opinion about the kconfig parts.

About the changes to the tests and duplicated files, I leave that to their respective owners and @nashif and @andrewboie, maybe they can point a nicer way of doing it.

File: boards/posix/native_posix/Kconfig.board
Comment (aescolar) [reply boards/posix/native_posix/Kconfig.board]: That proposal of mine above would probably mean replacing every current instance in this PR branch of BOARD_NATIVE_POSIX with something else (say NATIVE_POSIX_COMMON), and then leaving in this file on BOARD_NATIVE_POSIX and BOARD_NATIVE_POSIX_64. It may not be worth the trouble.

**Comment by npitre** [comment]
On Sat, 13 Jul 2019, Alberto Escolar wrote:

> +# This is pretty x86 centric. Some more generic method would be needed.
> 
> Yep. I think it would be nicer to add a new kconfig choice for the POSIX arch, something like `POSIX_ARCH_TARGET`
> which could start with `POSIX_ARCH_TARGET{_X86,_X86_64}` and default to `POSIX_ARCH_TARGET_X86` if not `64BIT` (or have 64 bit one be selected by `NATIVE_POSIX_64BIT`).

I think you are right, however I'd prefer if that was done separately. 
I'd want for this PR to concern itself with the addition of 
native_posix_64 only, and optimizations could be done at a later time.

**Comment by npitre** [comment]
On Sat, 13 Jul 2019, Alberto Escolar wrote:

> -config BOARD_NATIVE_POSIX
> -	bool "Native POSIX"
> +config BOARD_NATIVE_POSIX_32BIT
> +	bool "Native POSIX for 32-bit host"
>  	depends on SOC_POSIX
> 
> As this is common for both, would it make sense to move it to BOARD_NATIVE_POSIX?

This is the items pulled into the board selection menu. I'm not sure how 
it could be done differently. Or maybe I misunderstood what you're 
asking.

**Comment by npitre** [comment]
On Sat, 13 Jul 2019, Alberto Escolar wrote:

> aescolar commented on this pull request.
> 
> 
> 
> > @@ -1,7 +1,15 @@
>  # SPDX-License-Identifier: Apache-2.0
>  
> +# This is pretty x86 centric. Some more generic method would be needed.
> +if (CONFIG_64BIT)
> +	zephyr_compile_options(-m64)
> 
> This is a bit ugly, as you only need this to override cmake/compiler/host-gcc/target.cmake:22-25.
> 
> Instead would it make sense to only add `-m32` here if not CONFIG_64BIT and have those lines in cmake/compiler/host-gcc/target.cmake be conditional also to not CONFIG_64BIT ?

Honestly, I think this is rather confusing even without my patch. All 
I'm doing here is switching an unconditional -m32 for a conditional -m32 
or -m64. I'm sure this ought to be improved. But as I said before, it 
would be best to do this separately not to overload this PR with 
multiple purposes.

**Comment by npitre** [comment]
On Sat, 13 Jul 2019, Alberto Escolar wrote:

> As the USB test patch already made it to master, please rebase without those commits to avoid confusion

Done.

**Comment by marc-hb** [comment]
> Honestly, I think this is rather confusing even without my patch. 

Indeed.

I think  the confusion is mainly because the -m32/-mx32/-m64 Data Type flags have "dual citizenship": on one hand they're toolchain specific, on the other hand they're -march-itecture specific. To keep the confusion alive they seem compatible with most but not all toolchains _and_  they seem portable across many 64bits architectures too. 

So in Zephyr these Data Type flags can live in:
A.  `/Arch/**` country with some toolchain logic if ever needed, or they can live in:
C. `cmake/Compiler/**` with some architecture conditionals as seen in existing `{xtools,host-gcc}/target.cmake` files.
B. There's no plan B cause nothing starts with "B"

In an alternate universe, maybe toolchains should have combined the two flags in one, e.g: -march=atom-m32. Ship has sailed but as I suggested some time back in the host-gcc/target.cmake comment, option A. seems better to me for at least these reasons:
- (the majority of ?) 32 bits-only architectures don't care and don't need to be bothered by this, better not add noise in the toolchains they care about.
- `-mx32` (PR #9522) in particular seems... complicated and better cordoned in `arch/x86_64/` IMHO
- I couldn't find any toolchain that uses a different name for these flags (@mped-oticon, @gnuless?), so the hypothetical toolchain logic in arch/** files may never be needed. Compare to the ugly _and_ (!) duplicated architecture logic already in compiler/*/target.cmake
- there's a big toolchain abstraction effort going on and it can or could provide one layer of indirection *if* ever needed.
- By convention, plan A is always the best.


> All I'm doing here is switching an unconditional -m32 for a conditional -m32
or -m64.

Yes but as pointed out by @aescolar the current submission makes the current mixup of options A. and C. **worse**. Not just quantitatively but also qualitatively: passing both -m32 and -m64 on the same command line(s) and depending on the order. Ugh.

> I'm sure this ought to be improved. But as I said before, it
would be best to do this separately not to overload this PR with
multiple purposes.

Well... it's relatively frequent for big new features to expose old and latent problems and be blocked on other PRs required to cleanup these latent problems first. Note @aescolar offered a "almost constant entropy" solution instead. Maybe you don't want to let that offer pass :-)

My 3 cents.

File: arch/posix/CMakeLists.txt
Comment (pabigot) [reply arch/posix/CMakeLists.txt]: This was done, and does resolve the problem.

**Review by pabigot — CHANGES_REQUESTED** [review]
On the whole I really like what this PR is going to add.  By cloning a couple missed board configuration files in the display and gui samples I can finally do native tests using SDL on a 64-bit Debian system (Debian sdl packages don't work with multilib).

However, I've confirmed that this solution does indeed build with both `-m32` and `-m64` in the compiler flags.  As @aescolar noted `cmake/compiler/host-gcc/target.cmake` has:
```
# There's only one 64bits ARCH (actually: -mx32). Let's exclude it to
# avoid a confusing game of "who's last on the command line wins".
# Maybe the -m32/-miamcu FLAGS should all be next to -march= in the
# longer term?
if(NOT CONFIG_X86_64)
  string(PREPEND CMAKE_ASM_FLAGS             "-m32 ")
  string(PREPEND CMAKE_C_FLAGS               "-m32 ")
  string(PREPEND CMAKE_CXX_FLAGS             "-m32 ")
  string(PREPEND CMAKE_SHARED_LINKER_FLAGS   "-m32 ") # unused?
endif()
```
With this PR the comment is no longer correct.  These flags should be excluded for the new target as well, not put off to a later time that may never come.

**Comment by npitre** [comment]
On Sun, 14 Jul 2019, Marc wrote:

> I think  the confusion is mainly because the -m32/-mx32/-m64 Data Type flags have "dual citizenship": on one hand they're toolchain specific, on the other hand they're -march-itecture specific. To keep the confusion alive they seem compatible with most but not all toolchains _and_  they seem portable across many 64bits architectures too. 
> 
> So in Zephyr these Data Type flags can live in:
> A.  `/Arch/**` country with some toolchain logic if ever needed, or they can live in:
> C. `cmake/Compiler/**` with some architecture conditionals as seen in existing `{xtools,host-gcc}/target.cmake` files.
> B. There's no plan B cause nothing starts with "B"
> 
> - By convention, plan A is always the best.

The moment I remove -m32 from C then _both_ native-posix and 
native-posix-64 fail to build, even if -m32/-m64 are listed in A. It 
appears to be some kind of mystic CMake autoconf-like issue as none of 
the Zephyr code is even built at that point, and I can't find the actual 
error CMake is complaining about either. So I went with plan C.

File: boards/posix/native_posix/Kconfig.board
Comment (aescolar) [reply boards/posix/native_posix/Kconfig.board]: >>  +	bool "Native POSIX for 32-bit host"
>>  	depends on SOC_POSIX
>>
>> As this is common for both, would it make sense to move it to BOARD_NATIVE_POSIX?
>
>This is the items pulled into the board selection menu. I'm not sure how
>it could be done differently. Or maybe I misunderstood what you're
>asking.

(it was just a nit) I meant the `depends on SOC_POSIX` line. Could it not be moved into the common BOARD_NATIVE_POSIX config.

**Comment by aescolar** [comment]
@npitre Note that CI is failing with this latest patch, you can reproduce locally with `sanitycheck -p native_posix -T samples/application_development/external_lib`

**Comment by aescolar** [comment]
@marc-hb  @gnuless @nashif @andrewboie @npitre 
Wouldn't this be satisfactory?
```diff
diff --git a/arch/posix/CMakeLists.txt b/arch/posix/CMakeLists.txt
index 82fc5e8cb3..e1ddf5b19e 100644
--- a/arch/posix/CMakeLists.txt
+++ b/arch/posix/CMakeLists.txt
@@ -1,7 +1,13 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # some gcc versions complain without this
-zephyr_compile_options_ifdef(CONFIG_64BIT -fPIC)
+
+if (CONFIG_64BIT)
+       zephyr_compile_options(-fPIC)
+else ()
+       zephyr_compile_options(-m32)
+       zephyr_ld_options(-m32)
+endif ()
 
 zephyr_compile_options(
   -MMD
diff --git a/cmake/compiler/host-gcc/target.cmake b/cmake/compiler/host-gcc/target.cmake
index 2e709f3281..79dab0b2cc 100644
--- a/cmake/compiler/host-gcc/target.cmake
+++ b/cmake/compiler/host-gcc/target.cmake
@@ -16,12 +16,7 @@ find_program(CMAKE_GDB          gdb    )
 
 # Maybe the -m32/-miamcu FLAGS should all be next to -march= in the
 # longer term?
-if(CONFIG_64BIT)
-  string(PREPEND CMAKE_ASM_FLAGS             "-m64 ")
-  string(PREPEND CMAKE_C_FLAGS               "-m64 ")
-  string(PREPEND CMAKE_CXX_FLAGS             "-m64 ")
-  string(PREPEND CMAKE_SHARED_LINKER_FLAGS   "-m64 ") # unused?
-elseif(NOT CONFIG_X86_64)
+if (CONFIG_X86)
   string(PREPEND CMAKE_ASM_FLAGS             "-m32 ")
   string(PREPEND CMAKE_C_FLAGS               "-m32 ")
   string(PREPEND CMAKE_CXX_FLAGS             "-m32 ")
@@ -45,7 +40,7 @@ find_program(CMAKE_CXX_COMPILER ${cplusplus_compiler}     CACHE INTERNAL " " FOR
 # The x32 version of libgcc is usually not available (can't trust gcc
 # -mx32 --print-libgcc-file-name) so don't fail to build for something
 # that is currently not needed. See comments in compiler/gcc/target.cmake
-if (NOT CONFIG_X86_64)
+if (CONFIG_X86)
   # Convert to list as cmake Modules/*.cmake do it
   STRING(REGEX REPLACE " +" ";" PRINT_LIBGCC_ARGS ${CMAKE_C_FLAGS})
   # This libgcc code is partially duplicated in compiler/*/target.cmake
```

**Comment by gnuless** [comment]
Maybe we should really rename `CONFIG_64BIT` to something like `CONFIG_LP64` so it more clearly reflects what it means.  (I also don't like the fact that the option itself starts with a number, and I'm surprised we don't run into problems with that..)

**Comment by aescolar** [comment]
> Maybe we should really rename 

That would be for another PR ;).
My question to you guys, was if a change like that would work for all uses you know of this `cmake/compiler/host-gcc/target.cmake` file. And I ask you, because you are the ones playing with all kinds of combinations of compilers and targets for x86 and x86_64

File: arch/posix/CMakeLists.txt
Comment (aescolar) [reply arch/posix/CMakeLists.txt]: Postponed

**Comment by marc-hb** [comment]
> it's relatively frequent for big new features to expose old and latent problems

That was a severe understatement. Grab some pop-corn and check PR #17518

**Comment by npitre** [comment]
On Mon, 15 Jul 2019, Alberto Escolar wrote:

> >This is the items pulled into the board selection menu. I'm not sure how
> >it could be done differently. Or maybe I misunderstood what you're
> >asking.
> 
> (it was just a nit) I meant the `depends on SOC_POSIX` line. Could it not be moved into the common BOARD_NATIVE_POSIX config.

No. Those Kconfig entries are pulled (included) straight into the middle 
of a choice section (so they're actually choice options not standalone 
config prompts) and therefore each of them need to have their own 
dependencies.

**Comment by npitre** [comment]
On Mon, 15 Jul 2019, Alberto Escolar wrote:

> Wouldn't this be satisfactory?
> [...]

I applied something very similar to your suggestion and it does work, so 
I pushed it.

I'm not fully clear about what's going on in 
cmake/compiler/host-gcc/target.cmake so whatever works for you is OK 
with me there.

File: cmake/compiler/host-gcc/target.cmake
Comment (marc-hb) [inline cmake/compiler/host-gcc/target.cmake]: No arch besides X86 and native_posix needed -m32?

**Review by pabigot — APPROVED** [review]

File: cmake/compiler/host-gcc/target.cmake
Comment (aescolar) [reply cmake/compiler/host-gcc/target.cmake]: native_posix does, but I hope setting adding the compiler and linker flags directly from the arch cmake file would do it.

**Review by aescolar — APPROVED** [review]

File: include/sys/util.h
Comment (ulfalizer) [reply include/sys/util.h]: Dunno what this discussion is about exactly, but I'll throw out `ptrdiff_t` just in case.

File: boards/posix/native_posix/Kconfig.board
Comment (ulfalizer) [reply boards/posix/native_posix/Kconfig.board]: I'm fine with either one. `NATIVE_POSIX` matches the `-DBOARD` flag, and `NATIVE_POSIX_32` makes it clear that it's the 32-bit version. Different pros. Not sure which one is nicer.

Unrelated, but I think our board selection thing is pretty confusing. Since a bunch of other, non-Kconfig-related stuff is set when you select the board on the command line, it couldn't ever work to change th board within Kconfig. Maybe it shouldn't even have a prompt.

**Comment by pabigot** [comment]
I'm in favor of seeing this merged soon so we can start tracking down problems related to 64-bit targets.

I've discovered the CFB driver doesn't work with this platform.  The `cfb_font` structure is 24-bytes long, starting with an 8 byte `void *`.  The compiler walks the array believing that the offset between elements is 24, but the linker script appears to position the elements with a stride of 32 bytes.  There's a simple fix to the struct declaration that fixes this for CFB which I'll submit separately when I've done the necessary validation, but generically it suggests a problem with `common-rom.ld`.   (And @marc-hb is justified in his concerns wrt #17518; the hours I spent on that yesterday learning about that part of Zephyr showing value.)

I do not think that this should be fixed in this PR, but I do think we should be prepared to see this in other drivers that arrange their data into regions.

**Comment by npitre** [comment]
On Tue, 16 Jul 2019, Ulf Magnusson wrote:

> +/* Helper to get the difference between two pointers as an int */
> +#define PTR_DIFF(y, x) ((int)((intptr_t)(y) - (intptr_t)(x)))
> 
> Dunno what this discussion is about exactly, but I'll throw out `ptrdiff_t` just in case.

That part moved to #17518 and is no longer included here.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**
